### ISSUE TITLE
Support for object-level "record count" in S3

### ISSUE THREAD

**Issue by danyanos** [confluentinc/kafka-connect-storage-cloud#201]
Hi,

We were interested in contributing an enhancement that would add "record count" to all objects that land in S3. Traditionally our company performs a "row count" check when raw data enters a new datastore. The user landing the data provides an "expected" row count, and an internal framework compares the expected number of rows with the actual. Mismatches are used to alert producers of bad data so that some remediation steps can be taken. 

We've received requests from multiple users to add this "record count" to the objects that our Kafka Connect clusters are landing in S3. We've put together a proof of concept that uses S3 object-level tagging to add a "recordCount" tag to each object uploaded by the S3 Connector. Since we map each Kafka partition to one object in S3, we're also adding "startOffset" and "endOffset" tags to each object. We're interested in contributing this enhancement back to the community, but first wanted to check if this aligns with the roadmap for the S3 Connector and is something the community would support?

**Comment by cricket007** [comment]
If anything, you could extend somewhere in here to add in the record count of the file to the S3 key. 

https://github.com/confluentinc/kafka-connect-storage-cloud/blob/master/kafka-connect-s3/src/main/java/io/confluent/connect/s3/TopicPartitionWriter.java#L473-L488

To not break existing expectations, you can add a configuration flag like `include.record.count`: `true/false`

**Comment by kkonstantine** [comment]
Hi @danyanos 

this sounds definitely like a welcome contribution. I'd love to see a related PR. 
However, here, I'd like to discuss a bit whether addition record count as a tag vs as metadata header. 

To me, it seems that the latter might be slightly preferable for this type of information. For example, record count is associate with a specific object version and is immutable. A change in the count implies a different object version, as someone would expect. What do you think? Other than that, again I think it'll be useful, indeed. 

I wouldn't be in favor of adding this info to the key. Also, it doesn't need a config flag if we think it's useful and doesn't have side-effects.

**Comment by danyanos** [comment]
Hey @kkonstantine, 

I can definitely clean up what we have and put together a PR. I saw there were some guidelines on how to build the project in the FAQ, but was curious if there was any guidance/best practices on submitting PRs? 

In terms of metadata vs tags this is something that we explored a bit. Unless there's something I'm missing it appears the AWS Java SDK only allows you to specify metadata when you initiate a multipart upload. Since we won't know the final record count until we complete the multipart upload this didn't seem like a viable option.

**Comment by kevintdu** [comment]
Hi @kkonstantine

@danyanos and I implemented a feature to tag starting offset, ending offset, and record count for each s3 object created by the S3 Connector about a year to help us track data ingestion at an enterprise level. We didn't find until recently to contribute this back to Confluent, but here's the contributing PR: https://github.com/confluentinc/kafka-connect-storage-cloud/pull/291

This solution has been successful implemented and used in our application for about a year ago, across about ~4000 S3 connector jobs. Let us know if you have any feedback.

A couple specific responses to address the original conversation from when this issue was created:
1. We explored using object metadata over object tagging. Philosophically an immutable metadata would be better, but s3 multipart upload operations only allow for metadata attachment at the opening of the upload. We want to associate this information with the final result of the file. In addition, we found that tags are linked to the object versions. Uploading and overwriting a file with a different record count resulted in the tags attached only to that latest version, and does not impact the tags attached with a previous version.
2. We added a config flag in the event that a user's s3 write permissions did not extend to tagging.
